[possibly-undefined] False-positive with nested try block and finally

**Bug Report**
Small false-positive I noticed after #14114.
Seems to be a scoping issue with `finally`.

/CC: @ilinum 

**To Reproduce**

```python
try:
    try:
        msg = some_func()
    except Exception:
        raise

    print(msg)  # Name "msg" may be undefined  [possibly-undefined]

finally:
    pass

print(msg)  # <-- should probably be emitted here
```

The first print will always safely execute since the inner Try block raises. If it throws an uncaught exception, the print will be skipped to the `finally` block and only the second print will execute which would then raise an error itself as `msg` would be undefined.

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: `mypy 1.0.0+dev.96ac8b3e71c743b02ea4e3c84da0248659e40f82 (compiled: no)`
- Mypy command-line flags: `--enable-error-code possibly-undefined`
- Python version used: 3.11

here's a much simpler example that appears to be an exact duplicate:

```python
def f(x: int) -> None:
    try:
        if x >= 5:
            y = 2
        else:
            raise AssertionError('unreachable')
        print(y)
    finally:
        print('finally!')
```

```console
$ mypy t.py --enable-error-code=possibly-undefined
t.py:7: error: Name "y" may be undefined  [possibly-undefined]
Found 1 error in 1 file (checked 1 source file)
```

--- a/mypy/partially_defined.py
+++ b/mypy/partially_defined.py
@@ -206,11 +206,13 @@ class DefinedVariableTracker:
         # disable_branch_skip is used to disable skipping a branch due to a return/raise/etc. This is useful
         # in things like try/except/finally statements.
         self.disable_branch_skip = False
+        self.in_finally = False
 
     def copy(self) -> DefinedVariableTracker:
         result = DefinedVariableTracker()
         result.scopes = [s.copy() for s in self.scopes]
         result.disable_branch_skip = self.disable_branch_skip
+        result.in_finally = self.in_finally
         return result
 
     def _scope(self) -> Scope:
@@ -579,7 +581,9 @@ class PossiblyUndefinedVariableVisitor(ExtendedTraverserVisitor):
         self.tracker.end_branch_statement()
 
         if o.finally_body is not None:
+            self.tracker.in_finally = True
             o.finally_body.accept(self)
+            self.tracker.in_finally = False
 
     def visit_while_stmt(self, o: WhileStmt) -> None:
         o.expr.accept(self)
@@ -620,7 +624,10 @@ class PossiblyUndefinedVariableVisitor(ExtendedTraverserVisitor):
     def visit_name_expr(self, o: NameExpr) -> None:
         if o.name in self.builtins and self.tracker.in_scope(ScopeType.Global):
             return
-        if self.tracker.is_possibly_undefined(o.name):
+        if (
+            self.tracker.is_possibly_undefined(o.name)
+            and self.tracker.in_finally == self.tracker.disable_branch_skip
+        ):
             # A variable is only defined in some branches.
             self.variable_may_be_undefined(o.name, o)
             # We don't want to report the error on the same variable multiple times.

--- a/test-data/unit/check-possibly-undefined.test
+++ b/test-data/unit/check-possibly-undefined.test
@@ -717,7 +717,7 @@ def f7() -> int:
             raise BaseException
         return x  # No error.
     finally:
-        pass
+        y = x  # E: Name "x" may be undefined
     return 0
 [builtins fixtures/exception.pyi]
 

--- a/test-data/unit/check-possibly-undefined.test
+++ b/test-data/unit/check-possibly-undefined.test
@@ -708,6 +708,17 @@ def f6() -> int:
     finally:
         a = x  # E: Name "x" may be undefined
     return a
+
+def f7() -> int:
+    try:
+        if int():
+            x = 1
+        else:
+            raise BaseException
+        return x  # E: Name "x" may be undefined
+    finally:
+        pass
+    return 0
 [builtins fixtures/exception.pyi]
 
 [case testTryElse]

--- a/test-data/unit/check-possibly-undefined.test
+++ b/test-data/unit/check-possibly-undefined.test
@@ -678,7 +678,7 @@ def f3() -> int:
         pass
     finally:
         y = x  # E: Name "x" may be undefined
-    return x
+    return x  # E: Name "x" may be undefined
 
 def f4() -> int:
     try:
@@ -715,7 +715,7 @@ def f7() -> int:
             x = 1
         else:
             raise BaseException
-        return x  # E: Name "x" may be undefined
+        return x  # No error.
     finally:
         pass
     return 0

--- a/test-data/unit/check-possibly-undefined.test
+++ b/test-data/unit/check-possibly-undefined.test
@@ -628,7 +628,7 @@ def f7() -> int:
             x = 1
         assert False
     except:
-        pass
+        y = x  # E: Name "x" may be undefined
     return x  # E: Name "x" may be undefined
 [builtins fixtures/exception.pyi]
 

